### DISCUSSION TITLE
The Mandelbrot Set!

### DISCUSSION THREAD

**Discussion by ProxyPlayerHD** [cc65/cc65#1689] [Show and tell]
For some reason i had the need to draw the Mandelbrot set using my 65C02 SBC, since it only has a Serial connection i had to draw it with ASCII Characters.
the code i used was a modified version from rosettacode ([Original](https://rosettacode.org/wiki/Mandelbrot_set#C)), not very efficent but easy to adapt to cc65.
of course one obvious change was from writing to a file, to printing to Serial.
This was also a great test for my FixedMath Library. (well atleast for fxmul and fxdiv)

My first image was only 80x25 characters large and used multiple different symbols to add a gradient to the whole thing:
<img width="614" alt="ttermpro_0" src="https://user-images.githubusercontent.com/13576843/155863154-181a1f34-cbd0-417b-beb6-6a16751c1282.png">
It took 78s (1.3min) to finish.

But then i noticed that TeraTerm (the Terminal Software i'm using) supports color!
so for the next image i used the amount of Iterations as a grayscale value for the printed character and just used the same character for the whole image. I also made the screen larger, so i can render the image at a higher resolution (200x64 characters):
<img width="1514" alt="ttermpro_1" src="https://user-images.githubusercontent.com/13576843/155863394-07b67b45-8639-4de4-a87a-5339aab7490b.png">
This one took 498s (8.3min) to finish.

and then i set the font size to 1 and made the window as large as i could fit it on my screen (1000x440 characters). I also inverted the colors because i thought it would look cooler (and it does, IMO):
<img width="1514" alt="ttermpro_2" src="https://user-images.githubusercontent.com/13576843/155863465-bc7f04ac-7610-4fab-a2d7-c149b2fdfb2b.png">
This beast took 16957s (4.7h) to finish, i went shopping for groceries while it was running. keep in mind that my 65C02 runs at 20MHz with no wait states on the RAM that contains the program. A C64 would likely sit for multiple days on this.

here the final image from me, where i used [this site](https://sciencedemos.org.uk/mandelbrot.php) to find a nice spot in the set and set CxMin/CxMax/CyMin/CyMax to what the site showed as Left/Right/Top/Bottom (in that order), 400x140 characters:
<img width="1214" alt="ttermpro_2022-02-27_03-49-28" src="https://user-images.githubusercontent.com/13576843/155866151-bbc77a1c-bf98-42b0-9d27-69078e00cb24.png">
This one took 4689s (1.3h) to finish.

and finally, here the whole code, i rewrote it a bit to use stdio's printf instead of my custom Serial functions, which should make it much more portable!
```c
#include <stdint.h>
#include <stdio.h>
#include "FixedMath.h"

// My SBC has a CPLD that is programmed to increment a 32-bit counter every 20000 clock cycles, which is exactly once every millisecond
uint32_t clock(void){
	return *((uint32_t*)0xF7FC);		// Reads the millisecond Counter
}

#define iXmax		80			// Terminal Width
#define iYmax		25			// Terminal Height
#define FIX_ONE		0x00010000		// 1.0 (fixed point)
#define CxMin		0xFFFE2000		// -1.875 (fixed point)
#define CxMax		0x0000E000		//  0.875 (fixed point)
#define CyMin		0xFFFEE000		// -1.125 (fixed point)
#define CyMax		0x00012000		//  1.125 (fixed point)
#define IterMax		255			// Maximum Iteration Count
#define ER		2			// Escape Radius
#define ESC2		((ER*ER)*FIX_ONE)	// Escape Radius ^ 2 (fixed point)

#define termcolor(X,Y,Z) printf("\x1B[38;2;");printf("%d", (uint8_t)X);printf(";");printf("%d", (uint8_t)Y);printf(";");printf("%d", (uint8_t)Z);printf("m")
#define termreset() printf("\x1B[39m")
#define cls() printf("\x1B[2J")

int main(){
	
	fix16_t Cx, Cy, Zx, Zy, Zx2, Zy2;
	uint16_t iX, iY;
	uint16_t Iteration;
	uint32_t StartTime = clock();					// Start measuring the time
	fix16_t PixelWidth = fxdiv((CxMax - CxMin), fxfromint(iXmax));
	fix16_t PixelHeight = fxdiv((CyMax - CyMin), fxfromint(iYmax));
	
	cls();		// Clear the Terminal Screen
	
	for (iY = 0; iY < iYmax; iY++){
		
		Cy = CyMin + fxmul(fxfromint(iY), PixelHeight);		// Cy = CyMin + (iY * PixelHeight)
		
		if (fxabs(Cy) < (PixelHeight >> 1)) Cy = 0;
		
		for (iX = 0; iX < iXmax; iX++){
			Cx = CxMin + fxmul(fxfromint(iX), PixelWidth);	// Cx = CxMin + (iX * PixelWidth)
			
			Zx = 0;
			Zy = 0;
			Zx2 = 0;	// fxmul(Zx, Zx)
			Zy2 = 0;	// fxmul(Zy, Zy)
			
			for (Iteration = 0; (Iteration < IterMax) && ((Zx2+Zy2) < ESC2); Iteration++){
				Zy = (fxmul(Zx, Zy) << 1) + Cy;		// Zy = (Zx * Zy * 2) + Cy
				Zx = Zx2 - Zy2 + Cx;			// Zx = Zx2 - Zy2 + Cx
				Zx2 = fxmul(Zx, Zx);			// Zx2 = Zx*Zx
				Zy2 = fxmul(Zy, Zy);			// Zy2 = Zy*Zy
			}
			termcolor(Iteration, Iteration, Iteration);	// Set the Brightness of the Character based on how many Iterations were done
			printf("#");					// the more Iterations, the darker the character that gets printed
			
			// If your Terminal doesn't support color, comment the 2 lines above and uncomment the lines below
			
			// if (Iteration == IterMax){
				// ftprintc('#');
			// }
			// else if(Iteration > 15){
				// ftprintc('*');
			// }
			// else if(Iteration > 5){
				// ftprintc('.');
			// }
			// else{
				// ftprintc(' ');
			// }
		}
		printf("\n");						// Everytime a row is done, print a single \n character
	}
	termreset();							// Reset the Terminal Color to default, if no Color is supported, just comment this line out
	printf("%lu", (clock() - StartTime) / 1000);			// Print how many seconds it took
	return 0;
}
```
compiled, this printf version is ~4kB large.
the code can very likely be optimized quite a bit, both for speed and size.
would be interesting to see if anyone wants to play around with this, maybe add some actual colors, use actual graphics instead of a terminal, or do other fractals.

overall i just wanted to share this because i was proud of it and thought it looked pretty cool.
anyway, it's like 4 AM here and should really go to bed.

**Comment by Andrew900460** [comment]
Cool!

Just curious, are you open to sharing your fixed point library? I already have my own set of functions I've made. But I'm simply curious about how you went about coding your functions.
I understand if you don't want to share though.

**Reply by ProxyPlayerHD** [reply]
i have already made the library public, you should be able to find it on my profile:

https://github.com/ProxyPlayerHD/cc65-FixedMath

it's basically just a straight port of the "libfixmath" library, except i removed all the conditional stuff and rewrote the multiplication and division functions by hand in assembly.

**Reply by Andrew900460** [reply]
ahhhhh! cool!
